IOPool/Output/src/RootOutputFile.cc

```diff
@@ -682,70 +684,53 @@ namespace edm {
                 StoredProductProvenanceVector* productProvenanceVecPtr,
-                ModuleCallingContext const* mcc) {
+                ProductProvenanceRetriever const* provRetriever) {
 
     std::vector<std::unique_ptr<WrapperBase> > dummies;
 
-    bool const fastCloning = (branchType == InEvent) && (whyNotFastClonable_ == FileBlock::CanFastClone);
-
     OutputItemList const& items = om_->selectedOutputItemList()[branchType];
 
-    std::set<StoredProductProvenance> provenanceToKeep;
+    bool const doProvenance = (productProvenanceVecPtr != nullptr) && (om_->dropMetaData() != PoolOutputModule::DropAll);
+    bool const keepProvenanceForPrior = doProvenance && om_->dropMetaData() != PoolOutputModule::DropPrior;
 
-    ProductProvenanceRetriever const* provRetriever = nullptr;
-    {
-      EventPrincipal const* eventPrincipal = dynamic_cast<EventPrincipal const*>(&principal);
-      if (eventPrincipal) {
-        provRetriever =eventPrincipal->productProvenanceRetrieverPtr();
-        assert(provRetriever);
-      }
-    }
-    
+    bool const fastCloning = (branchType == InEvent) && (whyNotFastClonable_ == FileBlock::CanFastClone);
+    std::set<StoredProductProvenance> provenanceToKeep;
+    //
     //If we are dropping some of the meta data we need to know
     // which BranchIDs were produced in this process because
     // we may be storing meta data for only those products
+    // We do this only for event products.
     std::set<BranchID> producedBranches;
-    if(om_->dropMetaData() != PoolOutputModule::DropNone) {
-      for (auto bd: principal.productRegistry().allBranchDescriptions()) {
-        if(bd->produced()) {
+    if(doProvenance && branchType == InEvent && om_->dropMetaData() != PoolOutputModule::DropNone) {
+      Service<ConstProductRegistry> preg;
+      for(auto bd : preg->allBranchDescriptions()) {
+        if(bd->produced() && bd->branchType() == InEvent) {
           producedBranches.insert(bd->branchID());
         }
       }
     }
 
-    // Loop over EDProduct branches, fill the provenance, and write the branch.
+    // Loop over EDProduct branches, possibly fill the provenance, and write the branch.
     for(auto const& item : items) {
 
       BranchID const& id = item.branchDescription_->branchID();
       branchesWithStoredHistory_.insert(id);
 
       bool produced = item.branchDescription_->produced();
-      bool keepProvenance = productProvenanceVecPtr != nullptr &&
-                            (om_->dropMetaData() == PoolOutputModule::DropNone ||
-                             om_->dropMetaData() == PoolOutputModule::DropDroppedPrior ||
-                            (om_->dropMetaData() == PoolOutputModule::DropPrior && produced));
-      bool getProd = (produced || !fastCloning ||
-         treePointers_[branchType]->uncloned(item.branchDescription_->branchName()));
+      bool getProd = (produced || !fastCloning || treePointers_[InEvent]->uncloned(item.branchDescription_->branchName()));
```

Comment:
@Dr15Jones Actually, there currently is such a guarantee, because if it is not InEvent, fastCloning will always be false and we never get there. However, the guarantee is based on an assumption that may not always be true in the future. So, the code would be more maintainable and more future proof if InEvent were replaced by branchType here, so I will do so,